stdlib/src/builtin/string.mojo

```diff
@@ -660,6 +666,87 @@ fn isprintable(c: UInt8) -> Bool:
 # ===----------------------------------------------------------------------=== #
 
 
+fn _utf8_byte_type(b: UInt8) -> UInt8:
+    if b < 0b1000_0000:
+        return 0
+    elif b < 0b1100_0000:
+        return 1  # is continuation byte
+    elif b < 0b1110_0000:
+        return 2  # is 2 byte long
+    elif b < 0b1111_0000:
+        return 3  # is 3 byte long
+    else:
+        return 4  # is 4 byte long
```

Comment:
> Could this be done by just counting the leading ones?

That seems like an interesting idea. I'm not sure which would be more performant though. I've googled a bit and it seems leading zeroes speed varies some by ISA. 

```mojo
from bit import countl_zero
return max(4, countl_zero(~b))
```
what I'm seeing here is 1 branch, 1 negation op (no idea what cost), countl_zero (better perf in RISC than x86, but both are fast) for all first bytes

```mojo
    if b < 0b1000_0000:
        return 0
    elif b < 0b1100_0000:
        return 1  # is continuation byte
    elif b < 0b1110_0000:
        return 2  # is 2 byte long
    elif b < 0b1111_0000:
        return 3  # is 3 byte long
    else:
        return 4  # is 4 byte long
```
most ASCII alpabets will do just 1 operation, comparing in the first if and returning 0. The second most common will be a continuation byte since it is 1 for 2 byte sequences, 2 for 3 byte and 3 for 4 byte. But I have no idea what the cost is of this code having 5 branches.

Maybe `max(4, countl_zero(~b))` could be rewritten as `(countl_zero(~b) ^ 0b100 - 0b100) & 0b0111_1111` any idea what perf diff this would make or am I overcomplicating this?

Reply:
I'm not too worried about the perf implications and seems difficult to predict the real performance on a modern out-of-order machine. I suggest doing this mostly to make the code shorter and more expressive. But since you suggested using masks, a branchless solution is probably nice and I'm confident it will be performant enough:
```mojo
# note that both compilers and HW are really efficient at these bitwise logic
countl_zero(~(b & 0b1111_0000))
```